### ISSUE TITLE
panic: freepages: failed to get all reachable pages (page 3: multiple references)

### ISSUE THREAD

**Issue by JohnnyQQQQ** [nats-io/nats-streaming-server#769]
While testing our cluster for resiliency the following error occurred after recovering a server from a hard shutdown. It only occures sometimes, so it's probably an edgecase. The cluster is a 3 server cluster.

```log
[24] 2019/03/13 10:32:30.507372 [INF] STREAM: Starting nats-streaming-server[test] version 0.12.0
[24] 2019/03/13 10:32:30.507483 [INF] STREAM: ServerID: eKoq0U8jT54IzQK7kYoeZm
[24] 2019/03/13 10:32:30.507486 [INF] STREAM: Go version: go1.11.5
[24] 2019/03/13 10:32:30.512163 [INF] Starting nats-server version 1.4.1
[24] 2019/03/13 10:32:30.512198 [INF] Git commit [not set]
[24] 2019/03/13 10:32:30.512305 [INF] Starting http monitor on 0.0.0.0:8222
[24] 2019/03/13 10:32:30.512370 [INF] Listening for client connections on 0.0.0.0:4222
[24] 2019/03/13 10:32:30.512381 [INF] Server is ready
[24] 2019/03/13 10:32:30.512588 [INF] Listening for route connections on 0.0.0.0:6222
[24] 2019/03/13 10:32:30.514786 [INF] 10.244.9.154:6222 - rid:1 - Route connection created
[24] 2019/03/13 10:32:30.538772 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false false _NSS-test-send false false false <nil> true -1 250ms 2s 30s 0s 2m0s 2 0x954120 0x954170 0x9540d0 <nil> 0x954060 -1 8192    <nil> <nil> false}
[24] 2019/03/13 10:32:30.539713 [INF] 10.244.3.13:6222 - rid:3 - Route connection created
[24] 2019/03/13 10:32:30.539927 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false false _NSS-test-general false false false <nil> true -1 250ms 2s 30s 0s 2m0s 2 0x954120 0x954170 0x9540d0 <nil> 0x954060 -1 8192    <nil> <nil> false}
[24] 2019/03/13 10:32:30.540400 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false false _NSS-test-acks false false false <nil> true -1 250ms 2s 30s 0s 2m0s 2 0x954120 0x954170 0x9540d0 <nil> 0x954060 -1 8192    <nil> <nil> false}
[24] 2019/03/13 10:32:30.540878 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false false _NSS-test-raft false false false <nil> true -1 250ms 2s 30s 0s 2m0s 2 0x954120 0x954170 0x9540d0 <nil> 0x954060 -1 8192    <nil> <nil> false}
[24] 2019/03/13 10:32:30.541231 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false false _NSS-test-raft_snap false false false <nil> true -1 250ms 2s 30s 0s 2m0s 2 0x954120 0x954170 0x9540d0 <nil> 0x954060 -1 8192    <nil> <nil> false}
[24] 2019/03/13 10:32:30.541633 [INF] STREAM: Recovering the state...
[24] 2019/03/13 10:32:30.556203 [INF] STREAM: Recovered 2 channel(s)
[24] 2019/03/13 10:32:30.556425 [INF] STREAM: Cluster Node ID : nats-deployment-2.nats
[24] 2019/03/13 10:32:30.556444 [INF] STREAM: Cluster Log Path: /data/nats.io/cluster_log
panic: freepages: failed to get all reachable pages (page 3: multiple references)

goroutine 93 [running]:
github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt.(*DB).freepages.func2(0xc00032a2a0)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt/db.go:997 +0xfc
created by github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt.(*DB).freepages
	/go/src/github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt/db.go:995 +0x1b5
```

```config
port: 4222
cluster: {
  listen: 0.0.0.0:6222
  routes: ["nats://nats-deployment-0.nats:6222","nats://nats-deployment-1.nats:6222"]
}

# NATS Streaming specific configuration
streaming: {
  id: test
  store: file
  dir: /data/nats.io
  cluster: {
    node_id: nats-deployment-2.nats
    peers: [nats-deployment-0.nats,nats-deployment-1.nats]
    log_path: /data/nats.io/cluster_log
    raft_logging: true
  }
}
```

**Comment by JohnnyQQQQ** [comment]
We also got another panic (same kind, different place) from the same cluster in another round of chaos testing

```log
panic: freepages: failed to get all reachable pages (page 0: invalid type: unknown<00>)

goroutine 21 [running]:
github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt.(*DB).freepages.func2(0xc000088300)
	/go/src/github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt/db.go:997 +0xfc
created by github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt.(*DB).freepages
	/go/src/github.com/nats-io/nats-streaming-server/vendor/go.etcd.io/bbolt/db.go:995 +0x1b5
```

**Comment by kozlovic** [comment]
@JohnnyQQQQ Thank you for the report. Could you clarify what you mean by "hard shutdown"? Do you mean a kill -9 or power off the machine, etc..? Also, was that node panic'ing or recovery the leader prior to that or a follower? Thanks!

**Comment by kozlovic** [comment]
I have been doing a bit more experimentation with one of the node storing its state to a thumb drive and having a constant flow of published message. If I remove the thumb (without proper eject), I will get a panic (like in #754). There is a fix in progress from boltdb, but even with that, you may get some other kind of panic.
After putting back the thumb drive and restarting this node, I did get in some instances the exact panic that you posted here.

My take away is this: to minimize those risks you could run with:
```
file {
  sync: true
}
cluster {
  ...
  sync: true
}
```
(or command line `-file_sync=true -cluster_sync=true`), but be prepared in an enormous drop in performance (and all nodes need to have this setting otherwise one of the node will keep falling behind).

Alternatively, the idea of a cluster and replication is that you don't expect all nodes to fail at the same time. So if one fails, and fails to restart as in your case, you could remove its store (streaming and raft) and restart. It should be able to get all state back from the running nodes.

**Comment by JohnnyQQQQ** [comment]
> Could you clarify what you mean by "hard shutdown"?

It was a force shutdown in kubernetes `kubectl delete pod --force --grace-period 0 nats-deployment-2`

> Also, was that node panic'ing or recovery the leader prior to that or a follower?

Unfortunately I did overlook this detail, I will report back when I run the experiment again


We changes the cloud provider and therefore the storage engine. During the encountered panics we used rook.io with ceph. Now we use cloud provider specific storage and the panics comming from nats also dissapeared during our expiriments with chaostoolkit. So I think it was also related in some way to rook.io and/or ceph.

Feel free to close this issue.
